rfcs/splittable-prng.md

```diff
@@ -0,0 +1,103 @@
+
+(For a previous discussion of this problem where the suggestion to move to splittable PRNGs emerged, see  https://github.com/ocaml-multicore/ocaml-multicore/pull/582#pullrequestreview-683676282 )
+
+
+## PRINGO generators
+
+The [pringo](https://github.com/xavierleroy/pringo/) library provides two splittable PRNG implementations.
+
+| Name     | numeric type | speed wrt. Random | pure OCaml   | secure            | when to reseed?  |
+|----------|--------------|-------------------|--------------|-------------------|------------------|
+| Random   | int          | =                 | yes          | no                | ?                |
+| SplitMix | int64        | slightly faster   | no (C stubs) | no                | after 2^30 draws | 
+| ChaCha   | int32        | slightly slower   | no (C stubs) | yes (weak crypto) | after 2^64 draws |
+
+To give a sense of the performance difference, on my machine, on a micro-benchmarking drawing numbers in a tight loop (`make benchmarks` from the pringo directory):
+
+- drawing `int` has the same speed for all generators,
+- drawing `int64` takes 55% of the Random time with SplitMix and 93% of the Random time with ChaCha, and
+- drawing `float` values in [0;1] takes 66% of the Random time with SplitMix, and 120% of the Random time with ChaCha.
+
+Those differences are unlikely to be noticeable, as drawing random numbers is a neglectible part of most programs. Some very specific PRNG-bound algorithms (Monte-Carlo simulation, etc.) probably use custom PRNGs anyway.
+
+
+## Requirements for a standard-library PRNG
+
+Here is the list of potential requirements we considered to decide on including a PRNG implementation the standard library.
+
+### Non-concern: 32-bit hosts
+
+SplitMix, using 64-bit integers internally, is going to have lesser performance under 32-bit systems. We have more or less decided that we care less about the performance of 32-bit CPU architectures these days (we specifically discuss Javascript below), so we propose to not consider this in the choice process.
+
+### Concern: good randomness
+
+We want PRNGs that pass statistical tests for good randomness. All algorithms discussed here perform well under standard PRNG-quality test (diehard, etc.).
+
+
+### Possible concern: lesser portability of C stubs
+
+Random currently has a pure-OCaml implementation (except for the system-specific auto-seeding logic). Moving to C stubs may be an issue, at least for Mirage users, and require extra work from alternative implementations (js_of_ocaml, BuckleScript, etc.).
+
+SplitMix is a very simple algorithm, it is easy to provide a pure-OCaml version that should perform well. Porting Chacha, which is more elaborate, is more work, but it's also harder to predict performance for the pure-OCaml version.
```

Comment:
ChaCha reimplemented in OCaml will be awful.  The C implementation of parts of SplitMix is motivated by having decent performance with ocamlopt on 32-bit machines and with ocamlc (bytecode interpretation).  I agree that an OCaml implementation of SplitMix should be quite fast with ocamlopt on a 64-bit machine.

Reply:
I thought that you didn't care about 32-bit CPUs anymore? (If we keep x86-32 for cross-compilation we don't care about PRNG performance; riscv32 remains, and in general embedded targets.)

We could consider trying to have both a pure-OCaml and a C-stubs implementation, and choose between the two at configure-time. (Or by matching on Sys constants, but that would be less comfortable to Mirage people and other pure-OCaml constraints.)

Reply:
@xavierleroy as a non-expert, I'm curious as to how you can tell that pure-OCaml ChaCha performance would be bad. The [current implementation](https://github.com/xavierleroy/pringo/blob/bc32104/stubs.c#L172-L253) uses 15 mutable (stack-allocated) `uint32_t` variables; using `Int32.t ref` would introduce a lot of boxing.

But couldn't we use another approach, such that:
- unrolling the loop and using only (immutable) temporaries, and letting the register allocator sort it out,
- or using an `int32` bigarray stored in the generator state, which hopefully would let us read/write without boxing?

Reply:
I think an OCaml implementation of ChaCha would look pretty much like the OCaml implementation of MD5 that we use as a test (test/lib-digest/md5.ml) and that I used as a benchmark back in the days, and performance was not great compared with a pure C implementation.  The boxing of int32's in the state doesn't help.  The fact that ocamlopt doesn't recognize "rotate" instructions doesn't either.